File: src/main/java/com/paragon/client/systems/module/impl/misc/ChorusControl.java
package com.paragon.client.systems.module.impl.misc;

import com.paragon.api.event.network.PacketEvent;
import com.paragon.api.module.Category;
import com.paragon.api.module.Module;
import com.paragon.api.setting.Setting;
import com.paragon.api.util.render.RenderUtil;
import me.wolfsurge.cerauno.listener.Listener;
import net.minecraft.item.ItemChorusFruit;
import net.minecraft.network.play.client.CPacketConfirmTeleport;
import net.minecraft.network.play.client.CPacketPlayer;
import net.minecraft.network.play.client.CPacketPlayerTryUseItem;
import net.minecraft.network.play.server.SPacketPlayerPosLook;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.Vec3d;

import java.util.ArrayList;
import java.util.Collection;

/**
 * @author GentlemanMC, SooStrator1136
 */
public final class ChorusControl extends Module {

    private final Setting<Boolean> cPacketPlayer = new Setting<>("CPacketPlayer", true, null, null, null);
    private final Setting<Boolean> packetPlayerPosLook = new Setting<>("SPacketPlayerPosLook", true, null, null, null);

    private final Collection<CPacketPlayer> packets = new ArrayList<>(10);
    private final Collection<CPacketConfirmTeleport> teleportPackets = new ArrayList<>(2);

    private Vec3d renderPos = null; //Idk where to set this, your turn gentle

    private boolean ate = false;

    public ChorusControl() {
        super("ChorusControl", Category.MISC, "Cancels packets to let you not teleport");
    }

    @Override
    public void onDisable() {
        if (mc.getConnection() != null) {
            this.packets.forEach(mc.getConnection()::sendPacket);
            this.teleportPackets.forEach(mc.getConnection()::sendPacket);
        }

        this.packets.clear();
        this.teleportPackets.clear();
        this.ate = false;
    }

    @Override
    public void onRender3D() {
        if (this.renderPos == null) {
            return;
        }

        RenderUtil.drawNametagText("Player Chorus", this.renderPos, -1);
    }

    @Listener
    public void onPacketReceive(PacketEvent.PreReceive event) {
        if (event.getPacket() instanceof CPacketPlayerTryUseItem) {
            if ((((CPacketPlayerTryUseItem) event.getPacket()).getHand() == EnumHand.OFF_HAND ? mc.player.getHeldItemOffhand() : mc.player.getHeldItemMainhand()).getItem() instanceof ItemChorusFruit) {
                this.ate = true;
            }
        }
        if (this.ate) {
            if (event.getPacket() instanceof SPacketPlayerPosLook && this.packetPlayerPosLook.getValue()) {
                event.cancel();
            }

            if (event.getPacket() instanceof CPacketPlayer && this.cPacketPlayer.getValue()) {
                this.packets.add(((CPacketPlayer) event.getPacket()));
                event.cancel();
            }

            if (event.getPacket() instanceof CPacketConfirmTeleport) {
                this.teleportPackets.add(((CPacketConfirmTeleport) event.getPacket()));
                event.cancel();
            }
        }
    }

}

File: src/main/kotlin/com/paragon/impl/module/misc/ChorusControl.kt
package com.paragon.impl.module.misc

import com.paragon.impl.event.network.PacketEvent.PreReceive
import com.paragon.impl.module.Module
import com.paragon.impl.setting.Setting
import com.paragon.bus.listener.Listener
import com.paragon.impl.module.Category
import com.paragon.util.render.RenderUtil.drawNametagText
import net.minecraft.item.ItemChorusFruit
import net.minecraft.network.play.client.CPacketConfirmTeleport
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraft.network.play.client.CPacketPlayerTryUseItem
import net.minecraft.network.play.server.SPacketPlayerPosLook
import net.minecraft.util.EnumHand
import net.minecraft.util.math.Vec3d

/**
 * @author GentlemanMC, SooStrator1136
 */
object ChorusControl : Module("ChorusControl", Category.MISC, "Cancels packets to let you not teleport") {

    private val cPacketPlayer = Setting("CPacketPlayer", true) describedBy "Utilise CPacketPlayer packets"
    private val packetPlayerPosLook = Setting(
        "SPacketPlayerPosLook", true
    ) describedBy "Utilise SPacketPlayerPosLook packets"
    private val packets: MutableCollection<CPacketPlayer> = ArrayList(10)

    private val teleportPackets: MutableCollection<CPacketConfirmTeleport> = ArrayList(2)

    private val renderPos: Vec3d? = null //Idk where to set this, your turn gentle
    private var ate = false

    override fun onDisable() {
        if (minecraft.connection != null) {
            packets.forEach((minecraft.connection ?: return)::sendPacket)
            teleportPackets.forEach((minecraft.connection ?: return)::sendPacket)
        }

        packets.clear()
        teleportPackets.clear()
        ate = false
    }

    override fun onRender3D() {
        if (renderPos == null) {
            return
        }

        drawNametagText("Player Chorus", renderPos, -1)
    }

    @Listener
    fun onPacketReceive(event: PreReceive) {
        if (event.packet is CPacketPlayerTryUseItem) {
            if ((if (event.packet.hand == EnumHand.OFF_HAND) minecraft.player.heldItemOffhand else minecraft.player.heldItemMainhand).item is ItemChorusFruit) {
                ate = true
            }
        }

        if (ate) {
            if (event.packet is SPacketPlayerPosLook && packetPlayerPosLook.value) {
                event.cancel()
            }

            if (event.packet is CPacketPlayer && cPacketPlayer.value) {
                packets.add(event.packet)
                event.cancel()
            }

            if (event.packet is CPacketConfirmTeleport) {
                teleportPackets.add(event.packet)
                event.cancel()
            }
        }
    }
}